File: day1/day1.jl
function main()
  input = split(read("input", String), "\n\n") .|> (x -> split(x, "\n"))

  # MAP phase
  data = Vector{Int64}[]
  for x ∈ input
    push!(data, begin
      l = Int64[]
      for s ∈ x
        k = tryparse(Int64, s)
        k ≠ nothing && push!(l, k)
      end
      l
    end)
  end

  # REDUCE phase
  amount_calories = sum.(data)

  # Finding maximum
  n = length(amount_calories)
  max_ind = 1
  for i = 1:n
    if amount_calories[i] > amount_calories[max_ind]
      max_ind = i
    end
  end

  println(amount_calories[max_ind])

  # Finding second maximum
  snd_max_ind = 1
  for i = 1:n
    if amount_calories[i] > amount_calories[snd_max_ind] && i ≠ max_ind
      snd_max_ind = i
    end
  end

  # Finding last (third) maximum
  trd_max_ind = 1
  for i = 1:n
    if amount_calories[i] > amount_calories[trd_max_ind] && i ≠ max_ind && i ≠ snd_max_ind
      trd_max_ind = i
    end
  end

  println(sum(amount_calories[[max_ind, snd_max_ind, trd_max_ind]]))
end

main()

File: day1/day1.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func toList(s string) []int {
	nums := strings.Split(s, "\n")
	result := make([]int, 0)
	for _, n := range nums {
		k, err := strconv.Atoi(n)
		if err == nil { // just ignore parsing errors
			result = append(result, k)
		}
	}
	return result
}

func sum(arr []int) int {
	result := 0
	for _, n := range arr {
		result += n
	}
	return result
}

func main() {

	// --- Day 1: Calorie Counting ---

	data, err := os.ReadFile("./input")
	if err != nil {
		panic(err)
	}

	// MAP phase
	items_calories := make([][]int, 0)
	for _, v := range strings.Split(string(data), "\n\n") {
		l := toList(v)
		items_calories = append(items_calories, l)
	}

	// REDUCE phase
	amount_calories := make([]int, 0)
	for _, l := range items_calories {
		amount_calories = append(amount_calories, sum(l))
	}

	// finding the maximum's index
	ind_max := 0
	for i := range amount_calories {
		if amount_calories[i] > amount_calories[ind_max] {
			ind_max = i
		}
	}

	fmt.Printf("Part 1: %d\n", amount_calories[ind_max])

	// --- Part Two ---

	// need to fine the second with more calories
	snd_ind_max := 0
	for i := range amount_calories {
		if amount_calories[i] > amount_calories[snd_ind_max] && i != ind_max {
			snd_ind_max = i
		}
	}

	// now the last (third) one
	trd_ind_max := 0
	for i := range amount_calories {
		if amount_calories[i] > amount_calories[trd_ind_max] && i != ind_max && i != snd_ind_max {
			trd_ind_max = i
		}
	}

	result := amount_calories[ind_max] + amount_calories[snd_ind_max] + amount_calories[trd_ind_max]
	fmt.Printf("Part 2: %d\n", result)
}
